src/lib/pickles/README.md

```diff
@@ -0,0 +1,96 @@
+# Pickles - a modular typed inductive proof system
+
+Pickles is a framework that provides an interface for writing SNARK circuits as
+inductive rules, backed by a co-recursive kimchi verifier operating over the
+Pallas and Vesta curves.
+
+It is currently implemented to work with the polynomial IOP
+[PLONK](https://eprint.iacr.org/2019/953.pdf) and the recursion layer
+[Halo](https://eprint.iacr.org/2019/1021.pdf), generally called
+[Halo2](https://electriccoin.co/blog/explaining-halo-2/).
+It also provides an abstraction to deal with lookup arguments and custom gates compatible with PLONK.
+
+Pickles aims to be modular using the OCaml module system. Ideally, an inductive proof
+system can be parametrized by any finite field and 2-cycle of curves (called
+`Tick` and `Tock`). However, at the moment, it is hardcoded to be used with
+Pasta and Vesta curves. The code refers to the algebraic parameters as `Impls`
+and are passed to generic functions as a first class module.
+
+A particularity of Pickles is to use the power of the OCaml type system to
+encode runtime invariants like vector sizes, heteregenous list and others at
+the type level to avoid constructing some statements that might be wrong at
+compile time. Some encoded theories can be found in
+[`Pickles_types`](pickles_types).
+Some types are paremetrized by type-level defined naturals.
+
+## Coding guidelines
+
+Functions related to computations are parametrised by at least two types,
+suffixed respectively with `var` and `val`, which represent `in circuit` values
+and `out circuit` values. The type `[('var, 'val) t]` describes a mapping from
+the OCaml type `['val]` to a type representing the value using PlonK variables
+(`['var]`).
+Each in-circuit encoded values have a corresponding `'a Typ.t` value, which
+carries the in-circuit values, out-circuit values and the related circuit
+constraints.
+A nested module `Constant` should be defined to encode the out-circuit values
```

Comment:
Why "should"? Do we "need to"? or even must? Or is it something that is not enforced but that is "nice to have"?

Reply:
I changed to must. I hope I am not wrong.